desktop/src/main/java/bisq/desktop/main/offer/offerbook/OfferBook.java

```diff
@@ -83,44 +85,112 @@ public void onAdded(Offer offer) {
-                        log.warn("We had an old offer in the list with the same Offer ID. We remove the old one. " +
-                                "old offerBookListItem={}, new offerBookListItem={}", candidateWithSameId.get(), offerBookListItem);
-                        offerBookListItems.remove(candidateWithSameId.get());
+                    OfferBookListItem newOfferBookListItem = new OfferBookListItem(offer, hashOfPayload);
+                    removeDuplicateItem(newOfferBookListItem);
+                    offerBookListItems.add(newOfferBookListItem);  // Add replacement.
+                    if (log.isDebugEnabled()) {
+                        log.debug("onAdded: Added new offer {}\n"
+                                        + "\twith newItem.payloadHash: {}",
+                                offer.getId(),
+                                newOfferBookListItem.hashOfPayload == null ? "null" : newOfferBookListItem.hashOfPayload.getHex());
                     }
-
-                    offerBookListItems.add(offerBookListItem);
                 } else {
                     log.debug("We have the exact same offer already in our list and ignore the onAdded call. ID={}", offer.getId());
                 }
+                printOfferBookListItems("After onAdded");
             }
 
             @Override
-            public void onRemoved(Offer offer) {
-                removeOffer(offer, tradeManager);
+            public void onRemoved(Offer offer, P2PDataStorage.ByteArray hashOfPayload) {
+                printOfferBookListItems("Before onRemoved");
+                removeOffer(offer, hashOfPayload);
+                printOfferBookListItems("After onRemoved");
             }
         });
     }
 
-    public void removeOffer(Offer offer, TradeManager tradeManager) {
+    private void removeDuplicateItem(OfferBookListItem newOfferBookListItem) {
+        String offerId = newOfferBookListItem.getOffer().getId();
+        // We need to remove any view items with a matching offerId before
+        // a newOfferBookListItem is added to the view.
+        List<OfferBookListItem> duplicateItems = offerBookListItems.stream()
+                .filter(item -> item.getOffer().getId().equals(offerId))
+                .collect(Collectors.toList());
+        if (duplicateItems.size() > 0) {
+            duplicateItems.forEach(oldOfferItem -> {
+                offerBookListItems.remove(oldOfferItem);
+                if (log.isDebugEnabled()) {
+                    log.debug("onAdded: Removed old offer {}\n"
+                                    + "\twith payload hash {} from list.\n"
+                                    + "\tThis may make a subsequent onRemoved( {} ) call redundant.",
+                            offerId,
+                            oldOfferItem.getHashOfPayload() == null ? "null" : oldOfferItem.getHashOfPayload().getHex(),
+                            oldOfferItem.getOffer().getId());
+                }
```

Comment:
There are still a lot of debug logs. Do you want to keep them for the first release with these changes? If so, I think it would be good to add a TODO to remove them later as I don't think they're that valuable in the long run.

Reply:
RE comment below:  _Does this mean an item with null hash will be removed if only the id matches? Seems like a way to get around the hash checks._  

When a user clicks the view open, the list is populated by fillOfferBookListItems(), instantiating new OfferBookListItems with null hashes -- hashes are set only in onAdded or onRemoved.

When the OfferBook listener gets an onRemoved call, the hash match will fail if the list's match has a null hash, unless I do this extra null check.  If/when we create a transient OfferPayload hash field, fillOfferBookListItems() can load the view with non-null hashes, and the null hash checks can be removed.